Active Learning with AutoNLP & Prodigy (#190)
huggingface/blog: active-learning-with-autonlp-and-prodigy.md

Using Prodigy, we started labelling the dataset. When we had around 20 samples, we trained a model using AutoNLP. Prodigy doesn't export the data in AutoNLP format, so we wrote a quick and dirty script to convert the data into AutoNLP format:

```python
import json
import spacy

from prodigy.components.db import connect

db = connect()
prodigy_annotations = db.get_dataset("bbc")
examples = ((eg["text"], eg) for eg in prodigy_annotations)
nlp = spacy.blank("en")

dataset = []

for doc, eg in nlp.pipe(examples, as_tuples=True):
    try:
        doc.ents = [doc.char_span(s["start"], s["end"], s["label"]) for s in eg["spans"]]
        iob_tags = [f"{t.ent_iob_}-{t.ent_type_}" if t.ent_iob_ else "O" for t in doc]
        iob_tags = [t.strip("-") for t in iob_tags]
        tokens = [str(t) for t in doc]
        temp_data = {
            "tokens": tokens,
            "tags": iob_tags
        }
        dataset.append(temp_data)
    except:
        pass

with open('data.jsonl', 'w') as outfile:
    for entry in dataset:
        json.dump(entry, outfile)
        outfile.write('\n')
```

> This will provide us with a `JSONL` file which can be used for training a model using AutoNLP. The steps will be same as before except we will select `Token Classification` task when creating the AutoNLP project. Using the initial data we had, we trained a model using AutoNLP. The best model had an accuracy of around 86% with 0 precision and recall. We knew the model didn't learn anything. It's pretty obvious, we had only around 20 samples. 

Reviewer: You could also mention that you can version control the dataset inside a hf.co dataset repo, link to your sample dataset, and embed a screenshot of the dataset viewer interface inside the blog?

WDYT?

---
title: "Active Learning with AutoNLP and Prodigy"
thumbnail: /blog/assets/42_autonlp_prodigy/thumbnail.png
---

<h1>Active Learning with AutoNLP and Prodigy</h1>

<div class="blog-metadata">
    <small>Published Dec 17, 2021.</small>
    <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/master/active-learning-with-autonlp-and-prodigy.md">
        Update on GitHub
    </a>
</div>

<div class="author-card">
    <a href="/abhishek">
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/img_url?w=200&h=200&f=face" title="Gravatar">
        <div class="bfc">
            <code>abhishek</code>
            <span class="fullname">Abhishek Thakur</span>
        </div>
    </a>
</div>

> Active or online learning in the context of machine learning is a process in which you constantly update the data, train a model and serve it to the end user. It is an endless process and requires human interaction for labelling / creating the data. In this article, we will discuss how to use AutoNLP and Prodigy to build an active learning pipeline.

Reviewer 2: Some suggestion in wording, but feel free to take or leave what you like/don't like.

```suggestion
Active learning in the context of Machine Learning is a process in which you iteratively add labeled data, retrain a model and serve it to the end user. It is an endless process and requires human interaction for labeling/creating the data. In this article, we will discuss how to use [AutoNLP](https://huggingface.co/autonlp) and [Prodigy](https://prodi.gy/) to build an active learning pipeline.
```

## AutoNLP

> [AutoNLP](https://huggingface.co/autonlp) is a framework created by Hugging Face that helps you to build your own state-of-the-art deep learning models on your own dataset with almost no coding at all. AutoNLP is built on the giant shoulders of Hugging Face's [transformers](https://github.com/huggingface/transformers), [datasets](https://github.com/huggingface/datasets), [inference-api](https://huggingface.co/inference-api) and many other tools built by Hugging Face for internal and external purposes.

Reviewer 2: ```suggestion
[AutoNLP](https://huggingface.co/autonlp) is a framework created by Hugging Face that helps you to build your own state-of-the-art deep learning models on your own dataset with almost no coding at all. AutoNLP is built on the giant shoulders of Hugging Face's [transformers](https://github.com/huggingface/transformers), [datasets](https://github.com/huggingface/datasets), [inference-api](https://huggingface.co/inference-api) and many other tools.
```

With AutoNLP, you can train SOTA transformer models on your own custom dataset, fine-tune them (automatically) and serve them to the end-user. All models trained with AutoNLP are state-of-the-art and production-ready.

> At the time of writing this article, AutoNLP supports tasks like binary classification, regression, multi class classification, token classification (or NER task), question answering, summarization and more. You can find a list of all the supported tasks [here](https://huggingface.co/autonlp/). AutoNLP supports languages like English, French, German, Spanish, Hindi, Dutch, Swedish and many more. There is also support for custom models with custom tokenizers (in case your language is not supported by AutoNLP).

Reviewer 2: ```suggestion
At the time of writing this article, AutoNLP supports tasks like binary classification, regression, multi class classification, token classification (such as named entity recognition or part of speech), question answering, summarization and more. You can find a list of all the supported tasks [here](https://huggingface.co/autonlp/). AutoNLP supports languages like English, French, German, Spanish, Hindi, Dutch, Swedish and many more. There is also support for custom models with custom tokenizers (in case your language is not supported by AutoNLP).
```

## Prodigy

> [Prodigy](https://prodi.gy/) is an annotation tool developed by Explosion (the makers of Spacy). It is a web-based tool that allows you to annotate your data in real time and not just for Natural Language Processing tasks! Prodigy supports tasks like named entity recognition (NER), text classification, computer vision tasks and you can even create your own task. You can try the prodigy demo: [here](https://prodi.gy/demo).

Reviewer 2: ```suggestion
[Prodigy](https://prodi.gy/) is an annotation tool developed by Explosion (the makers of [spaCy](https://spacy.io/)). It is a web-based tool that allows you to annotate your data in real time and not just for Natural Language Processing tasks! Prodigy supports NLP tasks such as named entity recognition (NER) and text classification, but it's not limited to NLP! It  supports Computer Vision tasks and even creating your own tasks! You can try the Prodigy demo: [here](https://prodi.gy/demo).
```

Reviewer 3: I was just going to nitpick the spelling of "spaCy" but @osanseviero  beat me to it 😛 I like the edit suggestion!

> Please note that Prodigy is not free or open-source. It is a paid tool. You can find out more about it [here](https://prodi.gy/buy).

Reviewer 2: ```suggestion
Please note that Prodigy is not free nor open-source. It is a paid tool. You can find out more about it [here](https://prodi.gy/buy).
```

Reviewer 3: Maybe you could instead say something like "Note that Prodigy is a commercial tool" or similar? While it's true that it's not free open-source, it still follows the open-source spirit in the sense that you can view the code for all the built-in workflows and use it as a library to develop your own workflows aka. recipes with code.

Author: Sure. will take care of it

<img src="assets/42_autonlp_prodigy/data_view.png" width=500 height=250>

As we can see this is a classification dataset. There is a `Text` column which is the text of the news article and a `Category` column which is the class of the article. Overall, there are 5 different classes: `business`, `entertainment`, `politics`, `sport` & `tech`. 

Training a multi-class classification model on this dataset using AutoNLP is a piece of cake. 

Step 1: Download the dataset.

Step 2: Open [AutoNLP](https://ui.autonlp.huggingface.co/) and create a new project.

<img src="assets/42_autonlp_prodigy/autonlp_create_project.png">

Step 3: Upload the training dataset and choose auto-splitting.

<img src="assets/42_autonlp_prodigy/autonlp_data_multi_class.png">

Step 4: Accept the pricing and train your models.

<img src="assets/42_autonlp_prodigy/autonlp_estimate.png">

Please note that in the above example, we are training 15 different multi-class classification models. AutoNLP pricing can be as low as $10 per model. AutoNLP will select the best models and do hyperparameter tuning for you on its own. So, now, all we need to do is sit back, relax and wait for the results.

> After around 15mins, all models finished training and the results were ready. It seems like the best model scored 98.67% accuracy! 

Reviewer 2: ```suggestion
After around 15 minutes, all models finished training and the results are ready. It seems like the best model scored 98.67% accuracy! 
```

<img src="assets/42_autonlp_prodigy/autonlp_multi_class_results.png">

> So, we are now able to classify the articles in the dataset with an accuracy of 98.67%! But wait, we were talking about active learning and prodigy. What happened to those? We did use prodigy as we will see soon. We used prodigy to label this dataset for named entity recognition task. Before starting the labelling part, we thought it would be cool to have a project in which we are not only able to detect the entities in news articles but also categorize them. That's why we built this classification model on existing labels.

Reviewer 2: ```suggestion
So, we are now able to classify the articles in the dataset with an accuracy of 98.67%! But wait, we were talking about active learning and Prodigy. What happened to those? 🤔 We did use Prodigy as we will see soon. We used it to label this dataset for the named entity recognition task. Before starting the labeling part, we thought it would be cool to have a project in which we are not only able to detect the entities in news articles but also categorize them. That's why we built this classification model on existing labels.
```

## Active Learning

> The dataset we used did have categories but it didn't have labels for entity recognition. So, we decided to use Prodigy to label the dataset for antother task: named entity recognition or token classification.

Reviewer 2: ```suggestion
The dataset we used did have categories but it didn't have labels for entity recognition. So, we decided to use Prodigy to label the dataset for another task: named entity recognition.
```

> Once you have prodigy installed, you can simply run:

Reviewer 2: ```suggestion
Once you have Prodigy installed, you can simply run:
```

    $ prodigy ner.manual bbc blank:en BBC_News_Train.csv --label PERSON,ORG,PRODUCT,LOCATION

> Here `bbc` is the dataset that will be created by Prodigy. `blank:en` is the spacy tokenizer being used. `BBC_News_Train.csv` is the dataset that will be used for labelling. `PERSON,ORG,PRODUCT,LOCATION` is the list of labels that will be used for labelling.

Reviewer 2: ```suggestion
Let's look at the different values:
* `bbc` is the dataset that will be created by Prodigy. 
* `blank:en` is the `spaCy` tokenizer being used. 
* `BBC_News_Train.csv` is the dataset that will be used for labeling. 
* `PERSON,ORG,PRODUCT,LOCATION` is the list of labels that will be used for labeling.
```

> Once you run the above command, you can got to the prodigy web interface (usually at localhost:12000) and start labelling the dataset. Prodigy interface is very simple, intuitive and easy to use. The interface looks like the following:

Reviewer 2: ```suggestion
Once you run the above command, you can got to the Prodigy web interface (usually at localhost:12000) and start labeling the dataset. The Prodigy interface is very simple, intuitive and easy to use. It looks like the following:
```

Reviewer 3: By default, Prodigy should start the server at `localhost:8080` (but this can be customised via the `port` setting in the `prodigy.json` or the `PRODIGY_PORT` env variable).

Author: yes, you are right. i changed the port haha :D

<img src="assets/42_autonlp_prodigy/prodigy_ner.png">

> All you have to do is select which entity you want to label (PERSON, ORG, PRODUCT, LOCATION) and then select the text that belongs to the entity. Once you are done with one document, you can click on the green button and Prodigy will automatically provide you with next unlabelled document.

Reviewer 2: This is very cool! Maybe a gif could be useful here

Reviewer 3: Not sure if this is worth mentioning but a nice feature of the manual NER interface is that the selection "snaps" to the token boundaries: so you don't have to spend ages selecting all the individual characters perfectly, you can just roughly select the words. For single tokens, you can even double-click on them to select them.

Reviewer: i think it would make sense to mention! (or showcase in the GIF)

Author: @ines i did use the features you mentioned. do you have a gif for this specific feature? or shall i create something on my own?

Reviewer 3: There's a video [here](https://prodi.gy/docs/named-entity-recognition#manual-model) that shows it and that you could use for a gif (maybe just extract the most relevant seconds that show): https://prodi.gy/annotation_ner_make-gold-4273f5801e6c71aca958c63bd901e509.mp4

This video (starting at around 10:35) also shows the "snapping" explicitly and it's maybe a cleaner example: https://youtu.be/59BKHO_xBPA?t=635

But I think it could also be nice to show it on one of the specific examples from your blog post.

        iob_tags = [t.strip("-") for t in iob_tags]
        tokens = [str(t) for t in doc]
        temp_data = {
            "tokens": tokens,
            "tags": iob_tags
        }
        dataset.append(temp_data)
    except:
        pass

with open('data.jsonl', 'w') as outfile:
    for entry in dataset:
        json.dump(entry, outfile)
        outfile.write('\n')
```

This will provide us with a `JSONL` file which can be used for training a model using AutoNLP. The steps will be same as before except we will select `Token Classification` task when creating the AutoNLP project. Using the initial data we had, we trained a model using AutoNLP. The best model had an accuracy of around 86% with 0 precision and recall. We knew the model didn't learn anything. It's pretty obvious, we had only around 20 samples. 

After labelling around 70 samples, we started getting some results. The accuracy went up to 92%, precision was 0.52 and recall around 0.42. We were getting some results, but still not satisfactory. In the following image, we can see how this model performs on an unseen sample.

<img src="assets/42_autonlp_prodigy/a1.png">

> As you can see, the model is struggling. But it's much better than before. Previously, the model was not even able to predict anything in the same text. Atleast now, it's able to figure out that `Bruce` and `David` are names.

Reviewer 2: ```suggestion
As you can see, the model is struggling. But it's much better than before! Previously, the model was not even able to predict anything in the same text. At least now, it's able to figure out that `Bruce` and `David` are names.
```
